### ISSUE TITLE
Simple fee analytics

### ISSUE THREAD

**Issue by poliwop** [galacticcouncil/HydraDX-simulations#30]
Build out analysis with cadCAD simulations that plots LP profit/rate of return for different levels of fees, trade sizes.

This ticket need not involve sophisticated trade or agent modeling, we can plug that in later as we see fit.

**Comment by poliwop** [comment]
I am reviewing these notebooks a bit, looks like in some of them I need to change the # of timesteps in two separate places to run quick tests... In "...feesize_valpool" I have to change it in the notebook *and* I also have to change the "trade_count" variable in model_initialization.py? Can we restructure somehow so that we only have to change one variable?

**Comment by poliwop** [comment]
alright I have not fully worked through everything but I believe I've found a couple issues...
- on line 15 of model_initialization.py, LP2 needs to be added to agent_d. LP2 is the LP who will be accruing fees when R1 is sold for R2, so LP2 is the agent you should be looking at to evaluate fee profits for those types of trades.
- in TestSwap_fees_vary_tradesize&feesize_valpool.ipynb, I see the line
"profit_sim0 = profit_results2['val_pool'].iloc[0] - profit_results3['val_pool'].iloc[0]"
I could not remember exactly how iloc behaved so I looked at the value of profit_sim0, and it does not look correct to me... maybe double check this? Something is definitely not right, we are seeing LP profits go *down* as fee goes up. The assets in the pool are going the right direction, and so is "val_pool", so I think it's the profit calculation that is fishy...

**Comment by poliwop** [comment]
it looks like the changes to main have not been merged down to this branch?

I think before I review, those need to be merged down + tested. Probably best to also do a pull request for your changes to main so I can put my review notes there.

**Comment by KrisParuch** [comment]
@poliwop: A comparison of branches 'feat_fee_analysis' & 'main' on my local machine shows that I am up to date with 'main'.
Looking at the recent [commits to main](https://github.com/galacticcouncil/HydraDX-simulations/commits/main) I only see one commit yesterday which is coming from your work with andy on the arbitrage notebooks.

There is a difference however to the branch 'mb_event_verification' & I could merge these changes into my branch, this would however indicate that the changes to 'mb_event_verification' have not been committed (or pushed) to main.

Afterwards I can make a pull request to main for your review.

**Comment by poliwop** [comment]
Your branch definitely does not have the fixes I implemented (compare swap_assets in omnipool_amm.py, for example... or look at the Swap spec).

Can you please make sure your local branches are synched with the remote branches? I'm guessing your local main branch is out of date?

**Comment by KrisParuch** [comment]
Yes, thanks for the hint. I was assuming that this happens automatically when I select 'update from main' in github desktop without knowing that this only updates from my _local_ version.

Should be good now.

I'll be pushing the updated notebooks with experiments results soon; need to get rid of one final initialization error first and they should be good to go. The rest is fine

**Comment by KrisParuch** [comment]
@poliwop I updated the notebooks & initializations and run some experiments with the initial conditions suggested by BSCI in slack.
Now would be a good point in time to look at the two notebooks: 
- /hydradx_update/Analysis_LP_Profitability.ipynb
/hydradx_update/Analysis_LP_Profitability_5x3.ipynb

I also have two questions re:
- proper trade volume (see slack): we should aim to adjust volumes for each asset pair in the simulation to resemble reality
- market pressure: in my experiments I don't seem to get the sell_rx_rx_factor right (the division of which prices is the right one?). I was trying to make sense of it looking at the plot of val_pool (see at bottom of the notebook), but I get some patterns I do not like. maybe we can discuss in the call later

**Comment by poliwop** [comment]
@KrisParuch I guess I will add comments here... really there should probably be a pull request, but I will do a review without one for now.

https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/Analysis_LP_Profitability.ipynb?short_path=9a062d5#L141

As we discussed, this should be flipped... Numerator should be quantity of Lerna, to get LRNA/TKN price.

**Comment by poliwop** [comment]
I'd strongly suggest giving the "profit_results" variables more meaningful names for readibility
https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/Analysis_LP_Profitability.ipynb?short_path=9a062d5#L419

**Comment by poliwop** [comment]
Although I think this is technically correct, you should probably refer to lerna1 (initial quantity of LRNA in asset 1 sub-pool) instead of agent_d['LP1']['omniR1']. What you need is the quantity of lerna1 and it isn't obvious in this sheet why it should be equal to the number of LP shares in the possession of LP1.
https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/Analysis_LP_Profitability.ipynb?short_path=9a062d5#L446

**Comment by poliwop** [comment]
Eventually we should write a style guide for the team, but try to avoid ever importing "*". Best to either import specific functions, or do something like
"from model import model_initialization as mi"
so that your scope does not have variables with names not stated anywhere in the code
https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/Analysis_LP_Profitability.ipynb?short_path=9a062d5#L191

**Comment by poliwop** [comment]
You pulled out most of the numbers to the notebook which I like, just have trade count and prob_dict left:
https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/model/model_initialization.py#L81

I don't know what to do with the trader setup, I guess once that starts mattering (when we are tracking arbitrage profits?) we should pull that out too.

**Comment by poliwop** [comment]
Spent some time trying to figure out why trade size makes so little difference to LP 2 profits... it's because the trade sizes are both big enough to completely drain the pool of the second token (in the notebook with trade sizes of 1k and 2k). The code looks correct.

**Comment by poliwop** [comment]
if you are going to incorporate some sense of actual duration of each timestep, might make sense to annualize the returns
https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/Analysis_LP_Profitability.ipynb?short_path=9a062d5#L446

**Comment by poliwop** [comment]
When I try to fix the price fraction issue, I run into issues running the notebook... I think I've reviewed all I can until that is fixed, so I'll pass it back to you for now

**Comment by KrisParuch** [comment]
Thanks for your comments.

When I correct the division of the prices I start running into overflow errors, I guess that now the correct trade volume & sizing for resp. assets becomes critical.

Could you have a look at this division as well? (Not sure if the direction is correct):
https://github.com/galacticcouncil/HydraDX-simulations/blob/08826a0f1c6589d58c819e866fd44f86c803f9a2/hydradx_update/model/model_initialization.py#L51
